drivers/sensor/ms5611/ms5611.c

```diff
@@ -0,0 +1,551 @@
+				MS5611_CMD_CONVERT_D1_OSR_4096;
+		data->osr[MS5611_OSR_PRES_IDX].resp_time =
+				MS5611_RES_TIME_OSR_4096_US;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int ms5611_temperature_osr_set(const struct device *dev,
+				      const struct sensor_value *val)
+{
+	struct ms5611_data *data = dev->data;
+
+	switch (val->val1) {
+	case MS5611_OSR_256:
+		data->osr[MS5611_OSR_TEMP_IDX].ratio = MS5611_OSR_256;
+		data->osr[MS5611_OSR_TEMP_IDX].read_cmd =
+				MS5611_CMD_CONVERT_D2_OSR_256;
+		data->osr[MS5611_OSR_TEMP_IDX].resp_time =
+				MS5611_RES_TIME_OSR_256_US;
+		break;
+	case MS5611_OSR_512:
+		data->osr[MS5611_OSR_TEMP_IDX].ratio = MS5611_OSR_512;
+		data->osr[MS5611_OSR_TEMP_IDX].read_cmd =
+				MS5611_CMD_CONVERT_D2_OSR_512;
+		data->osr[MS5611_OSR_TEMP_IDX].resp_time =
+				MS5611_RES_TIME_OSR_512_US;
+		break;
+	case MS5611_OSR_1024:
+		data->osr[MS5611_OSR_TEMP_IDX].ratio = MS5611_OSR_1024;
+		data->osr[MS5611_OSR_TEMP_IDX].read_cmd =
+				MS5611_CMD_CONVERT_D2_OSR_1024;
+		data->osr[MS5611_OSR_TEMP_IDX].resp_time =
+				MS5611_RES_TIME_OSR_1024_US;
+		break;
+	case MS5611_OSR_2048:
+		data->osr[MS5611_OSR_TEMP_IDX].ratio = MS5611_OSR_2048;
+		data->osr[MS5611_OSR_TEMP_IDX].read_cmd =
+				MS5611_CMD_CONVERT_D2_OSR_2048;
+		data->osr[MS5611_OSR_TEMP_IDX].resp_time =
+				MS5611_RES_TIME_OSR_2048_US;
+		break;
+	case MS5611_OSR_4096:
+		data->osr[MS5611_OSR_TEMP_IDX].ratio = MS5611_OSR_4096;
+		data->osr[MS5611_OSR_TEMP_IDX].read_cmd =
+				MS5611_CMD_CONVERT_D2_OSR_4096;
+		data->osr[MS5611_OSR_TEMP_IDX].resp_time =
+				MS5611_RES_TIME_OSR_4096_US;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int ms5611_attr_set(const struct device *dev,
+			   enum sensor_channel chan,
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
+{
+	int ret;
+
+	if (attr != SENSOR_ATTR_OVERSAMPLING) {
+		return -ENOTSUP;
+	}
+
+	switch (chan) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		ret = ms5611_temperature_osr_set(dev, val);
+		break;
+	case SENSOR_CHAN_PRESS:
+		ret = ms5611_pressure_osr_set(dev, val);
+		break;
+	case SENSOR_CHAN_ALL:
+		ret = ms5611_pressure_osr_set(dev, val);
+		if (ret != 0)
+			break;
+
+		ret = ms5611_temperature_osr_set(dev, val);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+	return ret;
+}
+
+static int ms5611_fetch_temp_and_press(const struct device *dev)
+{
+	const struct ms5611_config *cfg = dev->config;
+	struct ms5611_data *data = dev->data;
+	uint8_t temp_data[3], press_data[3];
+	uint32_t raw_temp, raw_press;
+	int ret;
+	int32_t dT, temp, t2, p;
+	uint8_t i2c_cmd;
+	int64_t off, off2, sens, sens2;
+
+	/* Request temperature conversion */
+	i2c_cmd = data->osr[MS5611_OSR_TEMP_IDX].read_cmd;
+	ret = i2c_write_dt(&cfg->i2c_bus, &i2c_cmd, sizeof(i2c_cmd));
+	if (ret < 0) {
+		LOG_ERR("Failed to request temperature conversion");
+		return ret;
+	}
+
+	/* Wait for conversion to be finished */
+	k_sleep(K_USEC(data->osr[MS5611_OSR_TEMP_IDX].resp_time));
```

Comment:
A couple of questions, what is the longest time period that this, combined with the temp reading for calibration could cause blocking on the calling thread? Are there any status bits/interrupt lines to guarantee you have a new reading?

Reply:
According to the datasheet: values provided in lines 49-54 of ms5611.c file are "Max" conversion values. Thus, I suppose these are the worst case scenario values.
Unfortunately there is no interrupt line and the only way to get the information "if conversion has completed" is to ask the chip in next command sequence. If something went wrong (for example conversion wasn't completed) then the result is 0. This is why I am checking if the result is different from 0 in lines 251 and 282.